File: Web/backend/src/main/java/com/ByTech/ByTech/ShoppingCart/controllers/ShoppingCartController.java
package com.ByTech.ByTech.ShoppingCart.controllers;

import com.ByTech.ByTech.ShoppingCart.models.ShoppingCart;
import com.ByTech.ByTech.ShoppingCart.services.ShoppingCartService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.annotation.*;
import com.ByTech.ByTech.Message.Message;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;

import java.util.List;

@RestController
@RequestMapping("/shoppingList")
@CrossOrigin(origins = {"http://localhost:3000","http://localhost:3001"})
public class ShoppingCartController {
    private final ShoppingCartService shoppingCartService;

    @Autowired
    public ShoppingCartController(ShoppingCartService shoppingCartService) {
        this.shoppingCartService = shoppingCartService;
    }

    @GetMapping()
    public ResponseEntity<List<ShoppingCart>> getListByClient() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        if (authentication != null && authentication.getPrincipal() instanceof UserDetails) {
            String userName = ((UserDetails) authentication.getPrincipal()).getUsername();
            return new ResponseEntity<>(this.shoppingCartService.getListByClient(userName), HttpStatus.OK);
        } else {
            return new ResponseEntity<>(HttpStatus.UNAUTHORIZED);
        }
    }

    @GetMapping("/count/{client_id}")
    public ResponseEntity<Long> countByClient(@PathVariable("client_id")Long id){
        return new ResponseEntity<>(this.shoppingCartService.getCountByClient(id),HttpStatus.OK);
    }

    @GetMapping("/user/{userName}")
    public ResponseEntity<List<ShoppingCart>> getShoppingCartByUser(@PathVariable("userName") String userName) {
        List<ShoppingCart> shoppingCart = this.shoppingCartService.getListByClient(userName);
        return new ResponseEntity<>(shoppingCart, HttpStatus.OK);
    }

    @PostMapping(path = "/addProductToList")
    public ResponseEntity<Message> addProduct(@RequestBody ShoppingCart shoppingCart,
                                              BindingResult bindingResult) {
        if (bindingResult.hasErrors())
            return new ResponseEntity<>(new Message("Revise los campos"), HttpStatus.BAD_REQUEST);

        return shoppingCartService.addProduct(shoppingCart);
    }

    @PutMapping("/updateProductQuantity/{id}")
    public ResponseEntity<Message> updateProductQuantity(
            @RequestBody ShoppingCart shoppingCart,
            @PathVariable("id") Long id) {
        ShoppingCart existingItem = this.shoppingCartService.findById(id);
        if (existingItem != null) {
            existingItem.setAmount(shoppingCart.getAmount());
            this.shoppingCartService.updateProduct(existingItem);
            return new ResponseEntity<>(new Message("Cantidad de producto actualizada"), HttpStatus.OK);
        } else {
            return new ResponseEntity<>(new Message("El producto no está en el carrito"), HttpStatus.NOT_FOUND);
        }
    }

    @DeleteMapping("/clean/{item_id}")
    public ResponseEntity<Message> removeProduct(@PathVariable("item_id")Long id){
        this.shoppingCartService.removeProduct(id);
        return new ResponseEntity<>(new Message("Eliminado"),HttpStatus.OK);
    }

    @PutMapping(path = "{id}")
    public ShoppingCart updateUserById(@RequestBody ShoppingCart request , @PathVariable("id") Long id){
        return this.shoppingCartService.updateByID(request, id);
    }



}

File: Movil/backend/ByTech_movil/src/main/kotlin/com/example/ByTech_movil/ShoppingCart/controllers/ShoppingCartController.kt
package com.example.ByTech_movil.ShoppingCart.controllers

import com.example.ByTech_movil.ShoppingCart.models.ShoppingCart
import com.example.ByTech_movil.ShoppingCart.services.ShoppingCartService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.validation.BindingResult
import org.springframework.web.bind.annotation.*
import com.example.ByTech_movil.Message.Message
import org.springframework.security.core.Authentication
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.core.userdetails.UserDetails
import java.util.List

@RestController
@RequestMapping("/shoppingList")
@CrossOrigin(origins = ["exp://192.168.0.247:8081"])
class ShoppingCartController @Autowired constructor(
    private val shoppingCartService: ShoppingCartService
) {

    @GetMapping
    fun listByClient(): ResponseEntity<List<ShoppingCart>> {
        val authentication: Authentication = SecurityContextHolder.getContext().authentication
        return if (authentication.principal is UserDetails) {
            val userName: String = (authentication.principal as UserDetails).username
            ResponseEntity(shoppingCartService.getListByClient(userName), HttpStatus.OK)
        } else {
            ResponseEntity(HttpStatus.UNAUTHORIZED)
        }
    }


    @GetMapping("/count/{client_id}")
    fun countByClient(@PathVariable("client_id") id: Long?): ResponseEntity<Long> {
        return ResponseEntity(shoppingCartService.getCountByClient(id), HttpStatus.OK)
    }

    @GetMapping("/user/{userName}")
    fun getShoppingCartByUser(@PathVariable("userName") userName: String?): ResponseEntity<List<ShoppingCart>> {
        val shoppingCart = shoppingCartService.getListByClient(userName)
        return ResponseEntity(shoppingCart, HttpStatus.OK)
    }

    @PostMapping(path= ["/addProductToList"])
    fun addProduct(
        @RequestBody shoppingCart: ShoppingCart,
        bindingResult: BindingResult
    ): ResponseEntity<Message> {
        if (bindingResult.hasErrors()) return ResponseEntity(Message("Revise los campos"),HttpStatus.BAD_REQUEST)

        return shoppingCartService.addProduct(shoppingCart)
    }

    @PutMapping("/updateProductQuantity/{id}")
    fun updateProductQuantity(
        @RequestBody shoppingCart: ShoppingCart,
        @PathVariable("id") id: Long?
    ): ResponseEntity<Message> {
        val existingItem = shoppingCartService.findById(id)
        return if (existingItem != null) {
            existingItem.amount = shoppingCart.amount
            shoppingCartService.updateProduct(existingItem)
            ResponseEntity(Message("Cantidad de producto actualizada"), HttpStatus.OK)
        } else {
            ResponseEntity(Message("El producto no está en el carrito"), HttpStatus.NOT_FOUND)
        }

    }

    @DeleteMapping("/clean/{item_id}")
    fun removeProduct(@PathVariable("item_id") id: Long?): ResponseEntity<Message> {
        shoppingCartService.removeProduct(id)
        return ResponseEntity(Message("Eliminado"), HttpStatus.OK)
    }

    @PutMapping(value = ["{id}"])

    fun updateUserById(
        @RequestBody request: ShoppingCart,
        @PathVariable("id") id: Long?
    ): ResponseEntity<Any> {
        val updated = shoppingCartService.updateByID(request, id)
        return if (updated != null) {
            ResponseEntity(updated, HttpStatus.OK)
        } else {
            ResponseEntity(Message("No se encontró el carrito con ese ID"), HttpStatus.NOT_FOUND)
        }
    }


}